### PR TITLE
Ceed - add message storing error handler

### PR DESCRIPTION
**Pull request by jeremylt**
This change cleans up our error handling for the Python interface. Now the C error messages get raised as exceptions in the Python interface.

### CODE REVIEW

File: tests/t005-ceed.c [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  if (!errmsg || !strcmp(errmsg, "No error message stored"))
+    printf("Unexpected error message received: \"%s\"\n", errmsg);
+
+  // Check error message reset
+  CeedGetErrorMessage(ceed, &errmsg);
+  if (strcmp(errmsg, "No error message stored"))
+    printf("Unexpected error message received: \"%s\"\n", errmsg);
+
+  // Cleanup
+  CeedVectorDestroy(&vec);
```
Comment (jeremylt) [inline tests/t005-ceed.c:39]: Note to self, free message here

**Comment by valeriabarra** [comment]
Was I supposed to see a different error message now if I tested this in our jupyter notebook tutorials? I still see the same 
image

**Comment by jeremylt** [comment]
This is what I get on my end:
image

**Comment by valeriabarra** [comment]
I wasn't testing the vector's tutorial. I was testing `tutorial-0`, where I just created a ceed with a resource that I don't have, e.g., `ceed = libceed.Ceed(/gpu/cuda/gen)`

**Comment by jeremylt** [comment]
Right. That is the one edge case. We don't have a successfully created `ceed`, so we haven't set an error handler yet.

**Comment by jeremylt** [comment]
How about this change @valeriabarra?

File: python/ceed.py [hunk 2]
```diff
@@ -41,7 +42,10 @@ def __init__(self, resource="/cpu/self"):
 
         # libCEED call
         resourceAscii = ffi.new("char[]", resource.encode("ascii"))
-        lib.CeedInit(resourceAscii, self._pointer)
+        os.environ["CEED_ERROR_HANDLER"] = "return"
```
Comment (jedbrown) [inline python/ceed.py:45]: What if
```
$ CEED_ERROR_HANDLER=abort gdb -ex r ./try.py
```

File: interface/ceed.c [hunk 3]
```diff
@@ -928,4 +859,127 @@ int CeedErrorImpl(Ceed ceed, const char *filename, int lineno, const char *func,
...
+// LCOV_EXCL_START
+int CeedErrorReturn(Ceed ceed, const char *filename, int lineno,
+                    const char *func, int ecode, const char *format,
+                    va_list args) {
+  return ecode;
+}
+// LCOV_EXCL_STOP
+
+/**
+  @brief Error handler that stores the error message for future use.
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply interface/ceed.c:884]: I made it separate in the specific case where we do not have a ceed context yet Calloc'ed to store the message.

File: include/ceed.h [hunk 4]
```diff
@@ -173,6 +173,21 @@ CEED_EXTERN int CeedErrorImpl(Ceed, const char *, int, const char *, int,
...
+                               const char *, va_list);
+CEED_EXTERN int CeedErrorAbort(Ceed, const char *, int, const char *, int,
+                               const char *, va_list);
+CEED_EXTERN int CeedErrorExit(Ceed, const char *, int, const char *, int,
+                              const char *, va_list);
+CEED_EXTERN int CeedSetErrorHandler(Ceed ceed,
+                                    int (*eh)(Ceed, const char *, int,
+                                        const char *, int, const char *,
+                                        va_list));
+CEED_EXTERN int CeedGetErrorMessage(Ceed, char **errmsg);
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply include/ceed.h]: Updated

File: interface/ceed.c [hunk 5]
```diff
@@ -928,4 +859,127 @@ int CeedErrorImpl(Ceed ceed, const char *filename, int lineno, const char *func,
...
+  char errmsg[CEED_MAX_RESOURCE_LEN], errmsgva[CEED_MAX_RESOURCE_LEN/2],
+       *errmsg_copy;
+  vsprintf(errmsgva, format, args);
+  sprintf(errmsg, "%s:%d in %s(): %s", filename, lineno, func, errmsgva);
+  size_t slen = strlen(errmsg) + 1;
+
+  // Copy message
+  CeedMalloc(slen, &errmsg_copy);
+  memcpy(errmsg_copy, errmsg, slen);
+  ceed->errmsg = errmsg_copy;
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply interface/ceed.c]: I rewrote the approach to use a statically allocated array.

File: interface/ceed.c [hunk 6]
```diff
@@ -919,13 +853,149 @@ int CeedErrorImpl(Ceed ceed, const char *filename, int lineno, const char *func,
...
+    CeedErrorStore(ceed->parent, filename, lineno, func, ecode, format, args);
+
+  // Build message
+  char errmsg[CEED_MAX_RESOURCE_LEN], errmsgva[CEED_MAX_RESOURCE_LEN/2];
+  vsprintf(errmsgva, format, args);
+  sprintf(errmsg, "%s:%d in %s(): %s", filename, lineno, func, errmsgva);
+  size_t slen = strlen(errmsg) + 1;
+
+  // Copy message
+  memcpy(ceed->errmsg, errmsg, slen);
```
Comment (jedbrown) [inline interface/ceed.c:906]: What if `slen` is longer than the buffer? I guess I'd make the buffer have some extra capacity. I think you can do something like this to complexity of buffer management.
```c
  len = snprintf(ceed->errmsg, buffsize, "%s:%d in %s(): ", ...);
  vsnprintf(ceed->errmsg + len, buffsize-len, format, args);
```

Comment (jeremylt) [reply interface/ceed.c]: Agreed. Its less code and safer code. Changed.

**Comment by jeremylt** [comment]
Will merge after CI

**Pull request merged**